### PR TITLE
feat: add auto-heal job to restart unhealthy containers

### LINKED ISSUE #1333: ⚡️ Feature: Restart UnhealthyContainers
### Feature Description

I would like to request a feature to restart unhealthy containers, currently i have a small script checking this but would be nice to have this within arcane

### Problem It Solves

To manually restart a container to get it healthy 

### Proposed Solution

check container status if unhealthy

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you be willing to work on this feature?

- [ ] Yes, I'd like to implement this feature
- [ ] I could help with parts of this feature
- [x] No, I'm just suggesting the feature

### PR DESCRIPTION
**Pull request by garrett-edwards**
## What This PR Implements

**Related issue**
Closes #1333

Adds a scheduled auto-heal job that monitors unhealthy containers and automatically restarts them, with full frontend UI, notification support across all 10 providers, and a test notification option in the provider settings dropdown.

## Related Issue

Closes #1333

### Backend — Auto-Heal Job
- New `auto_heal_job.go` with restart-loop protection (in-memory tracking, configurable max restarts + time window)
- 5 new settings: enabled toggle, check interval, excluded containers, max restarts, restart window
- Job metadata registered so it appears in the Jobs UI
- Unit tests for the restart-loop logic

### Backend — Notifications
- Auto-heal notification event type with per-provider send functions for all 10 providers
- Auto-heal test notification type (`auto-heal`) in both handler and service layers
- Handler and service-layer unit tests for notification test type validation

### Frontend
- Auto-heal toggle switch on the Jobs tab with expanded settings (max restarts, restart window, container exclusion list)
- Container exclusion checkbox list with search (same pattern as auto-update, separate state)
- Auto-heal notification event subscription toggle for all 10 built-in providers
- "Test Auto-Heal Notification" option in the test provider dropdown (both built-in and default menus)
- i18n strings for all new UI elements
- TypeScript types for settings, job schedules, and notification form values
- Zod validation schemas for all provider forms

## Testing Done

- [x] Development environment started: `./scripts/development/dev.sh start`
- [x] Frontend verified at http://localhost:3000
- [x] Backend verified at http://localhost:3552
- [x] Manual testing completed: Auto-heal toggle, settings persistence, container exclusion, notification event subscriptions, test notification dropdown
- [x] No linting errors (e.g., `just lint all`)
- [x] Backend tests pass: `just test backend`

## Checklist

- [x] This PR is **not** opened from my fork's `main` branch

## AI Tool Used (if applicable)

AI Tool: Claude Code
Assistance Level: Significant
What AI helped with: Frontend implementation (Svelte 5 UI components, notification provider wiring, i18n, TypeScript types), test notification backend wiring
I reviewed and edited all AI-generated output: Yes
I ran all required tests and manually verified changes: Yes

## Additional Context

auto-heal-notification-events
auto-heal-jobs-tab
auto-heal-event-toggle

**Disclaimer** Greptiles Reviews use AI, make sure to check over its work.

To better help train Greptile on our codebase, if the comment is useful and valid **Like** the comment, if its not helpful or invalid **Dislike**

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds a comprehensive auto-heal feature that automatically restarts unhealthy containers with restart-loop protection, full notification support across all 10 providers, and a complete frontend UI.

**Backend Implementation**
- New `auto_heal_job.go` with in-memory restart tracking using mutex-protected maps
- Restart-loop protection with configurable max restarts (default: 5) and time window (default: 30 minutes)
- Container exclusion support via comma-separated list
- Proper integration with event logging and notification system
- 5 new settings: enabled toggle, check interval (cron), excluded containers, max restarts, restart window

**Notification System**
- Added `NotificationEventAutoHeal` event type
- Implemented send functions for all 10 providers (Discord, Email, Telegram, Signal, Slack, Ntfy, Pushover, Gotify, Matrix, Generic)
- Test notification support with `auto-heal` test type
- Follows existing notification patterns consistently

**Frontend UI**
- Auto-heal toggle and settings on Jobs tab
- Max restarts and restart window inputs with validation
- Container exclusion checkbox list with search functionality
- Event subscription toggles for all notification providers
- Test notification option in provider dropdown menus
- All i18n strings added

**Code Quality**
- Comprehensive unit tests for restart-loop protection logic
- Clean architecture with proper service injection
- Follows Go and Svelte 5 best practices
- All changes follow existing codebase patterns
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

- This PR is safe to merge with minimal risk
- The implementation is thorough, well-tested, and follows existing patterns throughout the codebase. No security vulnerabilities, logic errors, or breaking changes detected. The restart-loop protection prevents infinite restart cycles, and all notification providers are properly integrated.
- No files require special attention
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| backend/pkg/scheduler/auto_heal_job.go | Implements auto-heal job with restart-loop protection using in-memory tracking, container exclusion, and configurable limits. Clean architecture with proper mutex usage. |
| backend/pkg/scheduler/auto_heal_job_test.go | Comprehensive unit tests covering restart-loop protection logic including edge cases for window expiry, mixed timestamps, and multiple containers. |
| backend/internal/services/notification_service.go | Adds `SendAutoHealNotification` and 10 provider-specific send functions following existing patterns. Test type support properly integrated. |
| types/meta/job_metadata.go | Registers `auto-heal` job metadata with correct category, prerequisites, and settings keys. Allows manual execution. |
| backend/internal/bootstrap/jobs_bootstrap.go | Integrates auto-heal job into scheduler with proper registration, settings callbacks, and schedule change handlers. |
| frontend/src/routes/(app)/environments/[id]/components/JobsTab.svelte | Adds auto-heal UI with toggle, settings inputs (max restarts, restart window), and container exclusion list with search. Properly uses Svelte 5 runes. |

</details>

</details>

<sub>Last reviewed commit: 026907d</sub>

### CODE REVIEW

**Comment by kmendell** [comment]
Overall looks good at a glance, did you want to add the frontend for it? otherwise i can try to add it later if you arent comfortable with svelte

**Comment by kmendell** [comment]
The greptile comment and the linter warning as well are both valid concerns.

**Comment by garrett-edwards** [comment]
> Overall looks good at a glance, did you want to add the frontend for it? otherwise i can try to add it later if you arent comfortable with svelte

Let me take a look at the front end code and I'll let you know 😅

**Comment by kmendell** [comment]
@greptileai

File: backend/pkg/scheduler/auto_heal_job.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+}
+
+func (j *AutoHealJob) Name() string {
+	return AutoHealJobName
+}
+
+func (j *AutoHealJob) Schedule(ctx context.Context) string {
+	schedule := j.settingsService.GetStringSetting(ctx, "autoHealInterval", "*/30 * * * * *")
+	if schedule == "" {
+		schedule = "*/30 * * * * *"
```
Comment (kmendell) [inline backend/pkg/scheduler/auto_heal_job.go:56]: the cron implementation we use supports the seconds filed, so this would fire every 30 seconds, is that intended? I know this is a "healing job" so makes sense to run often , but eveyry 30 seconds may be too often? im not sure in all honesty.

**Review by kmendell — APPROVED** [review]
